sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageExtensions.cs

```diff
@@ -0,0 +1,157 @@
+            AmqpMessage.Create(message.GetAmqpSequenceBody().Select(s => new AmqpSequence(s)));
+
+        private static AmqpMessage ToAmqpValueMessage(this ServiceBusMessage message) =>
+            AmqpMessage.Create(new AmqpValue { Value = message.GetAmqpValueBody() });
+
+        public static AmqpMessage ToAmqpMessage(this ServiceBusMessage message) =>
+            message.GetAmqpBodyType() switch
+            {
+                AmqpBodyType.Data => message.ToAmqpDataMessage(viaBody: false),
+                AmqpBodyType.Sequence => message.ToAmqpSequenceMessage(),
+                AmqpBodyType.Value => message.ToAmqpValueMessage(),
+                // Unspecified
+                _ => message.ToAmqpDataMessage(viaBody: true)
+            };
+
+        public static AmqpBodyType GetBodyType(this AmqpMessage message)
+        {
+            if ((message.BodyType & SectionFlag.Data) != 0 && message.DataBody != null)
+            {
+                return AmqpBodyType.Data;
+            }
+
+            if ((message.BodyType & SectionFlag.AmqpSequence) != 0 && message.SequenceBody != null)
+            {
+                return AmqpBodyType.Sequence;
+            }
+
+            if ((message.BodyType & SectionFlag.AmqpValue) != 0 && message.ValueBody.Value != null)
+            {
+                return AmqpBodyType.Value;
+            }
+
+            return AmqpBodyType.Unspecified;
+        }
+
+        private static byte[] GetByteArray(this Data data)
+        {
+            switch (data.Value)
+            {
+                case byte[] byteArray:
+                    return byteArray;
+                case ArraySegment<byte> arraySegment when arraySegment.Count == arraySegment.Array?.Length:
+                    return arraySegment.Array;
+                case ArraySegment<byte> arraySegment:
+                {
+                    var byteArray = new byte[arraySegment.Count];
+                    Array.ConstrainedCopy(
+                        sourceArray: arraySegment.Array,
+                        sourceIndex: arraySegment.Offset,
+                        destinationArray: byteArray,
+                        destinationIndex: 0,
+                        length: arraySegment.Count);
+                    return byteArray;
+                }
+                default:
+                    return null;
+            }
+        }
+
+        // Converts the Data elements into ReadOnlyMemory<byte>, removing the Data elements that do not contain byte arrays.
+        // Also returns via the out parameter the flattened collection of bytes.
+        // A majority of the time, dataBody will only contain 1 element.
+        // The method is optimized for this situation to return nothing and only give back the pre-existing array via the flattened parameter.
+        private static IEnumerable<ReadOnlyMemory<byte>> ConvertAndFlatten(this IEnumerable<Data> dataBody, out byte[] flattened)
```

Comment:
Please take a close look at this method as it should reduce the amount of memory created for 99% of received messages. Previously, we would always create a List and always ToArray that list. Now, the actual array instance from the AmqpMessage's DataBody is returned. So, no new arrays are allocated.

Reply:
One thing that kind of worries me is how big `flattened` can become? Specifically, can it reach the size of object that is qualified for LOH (85,000 or more bytes)?

Reply:
A message can be up to 256KB for standard tier and 1MB for premium, so yeah the body can be over 85,000 bytes.

Reply:
To add a bit more context to what Josh was saying, it definitely CAN but the potential that it WILL is entirely dependent on the application consuming the client library.  It is common to find both applications that send many smaller messages and ones that send fewer, larger messages,  as well as those somewhere in-between.    

It is also entirely possible that a single application is sending messages of different sizes within the same context, such as in CQRS scenarios where `commands` tend to be larger and less frequent while `events` tend to be smaller and more frequent.

Reply:
I think, for now, we won't worry about this situation, unless we have a good solution for it. Like, some other recommended structure that can be used. For now, this was how T1 worked, but I'm always open to better optimization. Personally, I don't know a solution, though.